### PR TITLE
Type hints

### PR DESCRIPTION
**Pull request by lushoBarlett**
Implements declaration type hints, also allows parsing some type keywords as identifiers (for type terms) and declares `array` as a possible type in typechecker.

### CODE REVIEW

**Comment by lushoBarlett** [comment]
While implementing this I realized there's duplication in `Declaration` and `DeclarationList` some times, and that caused me problems, because I had written code for one of the functions and not for the other.

File: src/metacheck.cpp [hunk 1]
```diff
@@ -144,6 +144,12 @@ void metacheck(TypedAST::ReturnStatement* ast, TypeChecker& tc) {
 
 void metacheck(TypedAST::Declaration* ast, TypeChecker& tc) {
 	ast->m_meta_type = tc.new_meta_var();
+
+	if (ast->m_type) {
+		metacheck(ast->m_type, tc);
+		assert(ast->m_type->m_meta_type != tc.meta_value() && "Type hint must not be a value");
```
Comment (SebastianMestre) [inline src/metacheck.cpp]: Typehint must be a monotype, actually

File: src/metacheck.cpp [hunk 2]
```diff
@@ -152,8 +158,13 @@ void metacheck(TypedAST::DeclarationList* ast, TypeChecker& tc) {
 	for (auto& decl : ast->m_declarations)
 		decl.m_meta_type = tc.new_meta_var();
 
-	for (auto& decl : ast->m_declarations)
+	for (auto& decl : ast->m_declarations) {
+		if (decl.m_type) {
+			metacheck(decl.m_type, tc);
+			assert(decl.m_type->m_meta_type != tc.meta_value() && "Type hint must not be a value");
```
Comment (SebastianMestre) [inline src/metacheck.cpp]: Typehint must be a monotype

File: src/typecheck.cpp [hunk 3]
```diff
@@ -57,6 +59,17 @@ void typecheck(TypedAST::Declaration* ast, TypeChecker& tc) {
 
 	ast->m_value_type = tc.new_var();
 
+	if (ast->m_type) {
+		if (ast->m_type->type() == TypedASTTag::MonoTypeHandle) {
```
Comment (SebastianMestre) [inline src/typecheck.cpp]: typehints must be monotypes. This should be an assertion, not a check.

File: src/typecheck.cpp [hunk 4]
```diff
@@ -57,6 +59,17 @@ void typecheck(TypedAST::Declaration* ast, TypeChecker& tc) {
...
+			auto handle = static_cast<TypedAST::MonoTypeHandle*>(ast->m_type);
+			tc.m_core.m_mono_core.unify(ast->m_value_type, handle->m_value);
+		}
+		else {
+			auto handle = static_cast<TypedAST::TypeFunctionHandle*>(ast->m_type);
+			tc.m_core.m_tf_core.unify(ast->m_value_type, handle->m_value);
```
Comment (SebastianMestre) [inline src/typecheck.cpp]: Pretty sure this is a bug. `ast->m_value_type` must always be a monotype.

File: src/typecheck.cpp [hunk 5]
```diff
@@ -336,6 +349,17 @@ void typecheck(TypedAST::DeclarationList* ast, TypeChecker& tc) {
 		for (int u : verts) {
 			auto decl = index_to_decl[u];
 
+			if (decl->m_type) {
+				if (decl->m_type->type() == TypedASTTag::MonoTypeHandle) {
```
Comment (SebastianMestre) [inline src/typecheck.cpp]: assert, don't check

File: src/typecheck.cpp [hunk 6]
```diff
@@ -336,6 +349,17 @@ void typecheck(TypedAST::DeclarationList* ast, TypeChecker& tc) {
...
+					auto handle = static_cast<TypedAST::MonoTypeHandle*>(decl->m_type);
+					tc.m_core.m_mono_core.unify(decl->m_value_type, handle->m_value);
+				}
+				else {
+					auto handle = static_cast<TypedAST::TypeFunctionHandle*>(decl->m_type);
+					tc.m_core.m_tf_core.unify(decl->m_value_type, handle->m_value);
```
Comment (SebastianMestre) [inline src/typecheck.cpp]: this case is a bug

File: src/ct_eval.cpp [hunk 7]
```diff
@@ -247,9 +250,11 @@ TypedAST::DeclarationList* ct_eval(
 
 	for (auto& decl : ast->m_declarations) {
 		int meta_type = tc.m_core.m_meta_core.find(decl.m_meta_type);
+		// TODO: ban type hints from type function and mono type declarations?
```
Comment (SebastianMestre) [inline src/ct_eval.cpp:253]: Yes. That's probably wise.

**Review by SebastianMestre — CHANGES_REQUESTED** [review]
I don't think we should allow type functions in hints. Only monotypes. We can support that sugar in the future.

File: src/metacheck.cpp (hunk 1 above)
Comment (lushoBarlett) [reply src/metacheck.cpp]: Good to know, I wasn't sure myself.

**Comment by lushoBarlett** [comment]
> I don't think we should allow type functions in hints; only monotypes. We can add support for more sugar in the future.

That does simplify things a lot, I wasn't sure so I included the type function case as well, will delete it now tho.

File: src/ct_eval.cpp [hunk 8]
```diff
@@ -250,14 +250,15 @@ TypedAST::DeclarationList* ct_eval(
 
 	for (auto& decl : ast->m_declarations) {
 		int meta_type = tc.m_core.m_meta_core.find(decl.m_meta_type);
-		// TODO: ban type hints from type function and mono type declarations?
 		if (meta_type == tc.meta_typefunc()) {
+			assert(!decl.m_type_hint && "type hint not allowed in type function declaration");
```
Comment (SebastianMestre) [inline src/ct_eval.cpp:254]: I feel like this check should be in metacheck, and not here. Thoughts?

File: src/metacheck.cpp [hunk 9]
```diff
@@ -147,7 +147,7 @@ void metacheck(TypedAST::Declaration* ast, TypeChecker& tc) {
 
 	if (ast->m_type_hint) {
 		metacheck(ast->m_type_hint, tc);
-		assert(ast->m_type_hint->m_meta_type != tc.meta_value() && "Type hint must not be a value");
+		assert(ast->m_type_hint->m_meta_type == tc.meta_monotype() && "Type hint must be a monotype");
```
Comment (SebastianMestre) [inline src/metacheck.cpp:150]: I'm not sure... maybe we should unify instead of check?

File: src/metacheck.cpp [hunk 10]
```diff
@@ -161,7 +161,7 @@ void metacheck(TypedAST::DeclarationList* ast, TypeChecker& tc) {
 	for (auto& decl : ast->m_declarations) {
 		if (decl.m_type_hint) {
 			metacheck(decl.m_type_hint, tc);
-			assert(decl.m_type_hint->m_meta_type != tc.meta_value() && "Type hint must not be a value");
+			assert(decl.m_type_hint->m_meta_type == tc.meta_monotype() && "Type hint must be a monotype");
```
Comment (SebastianMestre) [inline src/metacheck.cpp:164]: Maybe unify instead of check? idk

**Comment by SebastianMestre** [comment]
You mentioned how its still accepted things like `x : int = 10;`, even though you removed the special handling for bare type functions.

That seemed really weird, but it actually makes a lot of sense. It's because of a quirk in the parser: when we already know we are parsing a type term, we implicitly wrap identifiers in type terms with no arguments, which means we get some nice syntactic sugar for free!

**Review by SebastianMestre — APPROVED** [review]
Looks good! Seems to be working quite well!

I have a few doubts, but they are more about tidiness and robustness, so they are not really urgent.

File: src/ct_eval.cpp (hunk 8 above)
Comment (lushoBarlett) [reply src/ct_eval.cpp:254]: I thought so as well, but it seemed like much less work to put them there hahaha. Perhaps we will fix that when we clean up.

File: src/metacheck.cpp (hunk 9 above)
Comment (lushoBarlett) [reply src/metacheck.cpp:150]: Fair point. I still like to have an assert there for us when we debug and test.

**Pull request merged**